[Validator] Fix support of Enum to `ConstraintValidator::formatValue`

PhoneixS opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #50000
| License       | MIT

I have added a simple check in the `formatValue` to see if the object is an enum and in that case, return the name of the enum instead of the "object" string.

PhoneixS committed:
[Validator] Fix support of Enum to `ConstraintValidator::formatValue`

PhoneixS commented:
> Hey!
> 
> Thanks for your PR. You are targeting branch "6.3" but it seems your PR description refers to branch "6.2". Could you update the PR description or change target branch? This helps core maintainers a lot.
> 
> Cheers!
> 
> Carsonbot

Sorry, first time making a PR. I think it's changed now.

stof commented:
I'm wondering whether this should be done only when the `OBJECT_AS_STRING` flag is passed

nicolas-grekas commented on src/Symfony/Component/Validator/ConstraintValidator.php:
```diff
@@ -93,6 +93,10 @@ protected function formatValue(mixed $value, int $format = 0): string
             }
 
+            if ($value instanceof \UnitEnum) {
+                return $value->name;
```
I'd suggest going with `$value::class.'::'.$value->name`.
can you please add a test case about it?

PhoneixS replied:
I don't see why it is better to check it with your code `$value::class.'::'.$value->name`. I think it is more readable with `instanceof`. Is there a case in which it can fail?

nicolas-grekas replied:
I'm not commenting on the check but on the return value.
We should return the name of the class + the name of case, not only the former.
`$value::class.'::'.$value->name`

PhoneixS replied:
ohhh, I see it now. I will change it.

PhoneixS replied:
Mmm, with that change it's a bit too verbose as it returns the full qualified class name. Although as this is more of a developer thing than something which would be seen by clients it could be ok.

PhoneixS commented:
I have changed so enums are treated more like objects, if there isn't an OBJECT_TO_STRING flag they are displayed as "enum"; if the flag is present it returns the enum name.

I have also added both test cases (with or without the flag).

PhoneixS commented:
Tests passed but failed because of unrelated  `Error: KO src/Symfony/Component/HttpClient`.

nicolas-grekas commented on src/Symfony/Component/Validator/ConstraintValidator.php:
```diff
@@ -88,6 +88,14 @@ protected function formatValue(mixed $value, int $format = 0): string
             return $value->format('Y-m-d H:i:s');
         }
 
+        if (\is_object($value) && $value instanceof \UnitEnum) {
```
is_object check is not needed, instanceof in enough

PhoneixS replied:
Correct, I'll remove it.

nicolas-grekas commented on src/Symfony/Component/Validator/ConstraintValidator.php:
```diff
@@ -88,6 +88,14 @@ protected function formatValue(mixed $value, int $format = 0): string
 
+        if ($value instanceof \UnitEnum) {
+            if ($format & self::OBJECT_TO_STRING) {
+                return $value::class.'::'.$value->name;
```
On a closer look, my advice was wrong I think, since the message already contains the class of the enum.
Let's make this be `return 'case '.$value->name;` so that messages looks like this?
"This value should not be identical to Symfony\Component\Validator\Tests\TestEnum case FirstCase"

PhoneixS replied:
In my opinion, we shouldn't add the 'case ' text, as it would be better added in the message itself.

nicolas-grekas replied:
but how can the message know about this? or maybe we should change the type to make it report "enum" and keep this as is?

nicolas-grekas replied:
Or maybe formatTypeOf should be changed instead to return `$value::class.'::'.$value->name`?

PhoneixS replied:
> Or maybe formatTypeOf should be changed instead to return `$value::class.'::'.$value->name`?

I see your point, however, I think that the function should only return the type (the class is the type and is enough for debugging) the same way the formatValue should only return the value (and the name is the textual representation of the enum value). Making each method return one thing whatever is called for. This is why I initially only return the name and probably we could just return to my initial implementation of returning the name of the enum.

nicolas-grekas replied:
OK, let's do this.

nicolas-grekas commented on src/Symfony/Component/Validator/ConstraintValidator.php:
```diff
@@ -88,6 +88,14 @@ protected function formatValue(mixed $value, int $format = 0): string
         }
 
+        if ($value instanceof \UnitEnum) {
+            if ($format & self::OBJECT_TO_STRING) {
```
I would remove this check personally, I don't see its value.

PhoneixS replied:
The first it's to check if it's an enum instead of a generic object.  The second is to maintain the same behavior as with objects, if the flag is set you return a value if not set, you return the string of what it is, an "enum" in this case and "object" in the object case.

I think we should keep them to maintain a similar functionality as with objects.

nicolas-grekas replied:
I don't see the link with OBJECT_TO_STRING flag here. This is unrelated to calling `__toString`. This method is creating a representation of the value. "enum" is not accurate enough, independently from any flag.

PhoneixS replied:
It's true that it could be better to have another flag like `self::NAME_OF_ENUM` specific for the enums (the same way objects have `self::OBJECT_TO_STRING` and `self::PRETTY_DATE` for dates. As this is my first PR I wanted to change the less possible but I can add it if it fits.

nicolas-grekas replied:
No need for any flag IMHO.

PhoneixS replied:
So you say to always return the name? In my opinion is better to let the developer decide what they want. Maybe someone wants to hide the actual value because it's shown to the user but others prefer to see it for debugging proposed. Isn't it the same reasoning as for dates and objects?

nicolas-grekas replied:
We display strings and numbers, so why not enums? I see each flag as a complication. The class of the enum is already exposed, not saying which case is concerned just worsen DX to me, this is not something that devs want to opt-in for, nor disable.

PhoneixS replied:
Ok, I will remove the flag and return always the name

nicolas-grekas replied:
Thanks, and sorry for the confusion :)

nicolas-grekas commented:
Thank you @PhoneixS.

PhoneixS commented:
You’re welcome. It's always nice to return a little bit to a great project.
